OsmAnd-java/src/main/java/net/osmand/search/core/SearchCoreFactory.java

```diff
@@ -571,19 +565,158 @@ public boolean isCancelled() {
+									// 		'Remseck am Neckar' [R 405291], 0 street(s) size 0 bytes 48.91138, 9.22288 - 48.84634, 9.31304
+									// 48.869724 Lon 9.276302
+//									'Remseck am Neckar' [R 405291], 0 street(s) size 0 bytes 48.91138, 9.22288 - 48.84634, 9.31304
+//									System.out.printf("Search %s %.5f, %.5f - %.5f, %.5f\n", res.toString(),
+//											MapUtils.get31LongitudeX((int) bbox.left), MapUtils.get31LatitudeY((int) bbox.top),
+//											MapUtils.get31LongitudeX((int) bbox.right), MapUtils.get31LatitudeY((int) bbox.bottom));
+									for (City boundary : resArray) {
+										int[] bb = boundary.getBbox31();
+										if (bb == null) {
+											continue;
+										}
+//										if (boundary.getName().toLowerCase().startsWith("remseck")) {
+//											System.out.println(boundary);
+//										}
+										QuadRect boundBox = new QuadRect(bb[0], bb[1], bb[2], bb[3]);
+										if (!QuadRect.intersects(boundBox, bbox)) {
+											continue;
+										}
+										if (matchCity(res, phrase, boundary, true)) {
+										// FIXME
+//										if(phrase.getFullSearchPhrase().contains(boundary.getName())) {
+//											System.out.printf("Found %s %.5f, %.5f - %.5f, %.5f\n", boundary.getName(),
+//													MapUtils.get31LongitudeX(bb[0]), MapUtils.get31LatitudeY(bb[1]),
+//													MapUtils.get31LongitudeX(bb[2]), MapUtils.get31LatitudeY(bb[3]));
+											cityResult.localeName = boundary.getName();
+											newParentSearchResult = cityResult;
+											break;
+										}
+									}
+								}
+							}
+							subSearchApiOrPublish(phrase, resultMatcher, res, streetsApi, newParentSearchResult, true);
+						} else if (res.objectType == ObjectType.BOUNDARY ) {
+							// 6 Dorfstraße Remseck am Neckar
+							// 4 Hofäckerstraße Kernen im Remstal
+							// 4 Am Heuhaus Weinstadt
+							// 4241 Cook Hollow Road Woodhull
+							// 301 West Main Street Valley View   
+//							11601 Kelly Hill Road Pine City // TODO improve
+							// 8508 PA 61 Coal Township
+							if (matchCity(null, phrase, (City) res.object, true)) {
+//							if (phrase.getFullSearchPhrase().toLowerCase().contains(res.localeName.toLowerCase())) {
+//								System.out.println("SUB " + res.object + " " + res.localeName + " " + res.objectType);
+								subSearchApiOrPublish(phrase, resultMatcher, res, this);
+							}
 						} else {
 							SearchPhrase nphrase = subSearchApiOrPublish(phrase, resultMatcher, res, cityApi);
-							searchPoiInCity(nphrase, res, resultMatcher);
+							// FIXME check if subsearch successful don't call search by name inside 
+//							searchPoiInCity(nphrase, res, resultMatcher);
+//							if(phrase.getFullSearchPhrase().toLowerCase().contains(res.localeName.toLowerCase())) {
+							if (matchCity(null, phrase, (City) res.object, true)) {
+//								System.out.println("SUBCITY " + res.object + " " + res.localeName + " " + res.objectType);
+								subSearchApiOrPublish(phrase, resultMatcher, res, this);
```

Comment:
Can we improve by checking results from 
` subSearchApiOrPublish(phrase, resultMatcher, res, cityApi);` first and if it didn't find then only go to 
`subSearchApiOrPublish(phrase, resultMatcher, res, this);`